### DISCUSSION TITLE
Key Input?

### DISCUSSION THREAD

**Discussion by meoowe** [korlibs/korge#2051] [Q&A] [answered]
I have been wondering what the differences are between `input.keys.justPressed`, `input.keys.justReleased` and  `input.keys.pressing`. None of them are working for me so I can't figure out what I'm meant to be doing. I have been looking at [this](https://docs.korge.org/views/input/#keys-1) and have tried all of them.

**Comment by maik-mursall** [comment, marked as answer]
Heya 👋🏻 

`justPressed` -> This key has been pressed during this frame (Will only be true for the first frame this key is pressed)
`justReleased` -> This key has been released during this frame (Will only be true for the last frame this key is pressed / the frame the key is released)
`pressing` -> This key is still being pressed this frame (Will be true for every frame this key is pressed in)

For your method you need to use an `Updater` to be able to receive your key information every frame:
```kotlin
override suspend fun SContainer.sceneMain() {
    addUpdater {
        println("W was just pressed: ${input.keys.justPressed(Key.W)}")
    }
    // Your code
```
(`addUpdater` doesn't have to be called first)

You can follow more examples [here](https://docs.korge.org/views/input/#keys) :)

**Reply by meoowe** [reply]
oh, thanks!

**Comment by maik-mursall** [comment]
As an alternative you can use the `keys` shorthand as well:

```kotlin
override suspend fun SContainer.sceneMain() {
    keys {
        justDown(Key.W) {
            println("justDown")
        }

        down(Key.W) {
            println("down")
        }

        up(Key.W) {
            println("up")
        }
    }
```

**Comment by soywiz** [comment]
BTW Updated the documentation with your clarifications: https://github.com/korlibs/docs.korge.org/commit/823b51b4a050a7bd41840a1da9cbe7bbfc57cca6#diff-21c9012ed18f4e4026df8f6bb7d576ba1bf0cd7b057be3b9207150b43f1c00bb

Thanks!
